### DISCUSSION TITLE
Can't read `NEXT_PUBLIC` env on client side

### DISCUSSION THREAD

**Discussion by LikeDreamwalker** [vercel/next.js#70099] [Help]
### Summary

I am trying to read a special env params to make my app be suitable for multiple scenarios.
For example, I use a param like this:
`NEXT_PUBLIC_BIZ_ENV=global`
And in my codes I will do things like this:
```tsx
      {process.env.NEXT_PUBLIC_BIZ_ENV === "global" ? (
        <SvgIcon
          sx={{
            width: renderSize.VOWidth,
            height: renderSize.VOHeight,
          }}
        >
...
```
And after I build it with docker, seems like only the server components, which will actually be rendered on server node env, can read the params right. For example, in this code I will try to login on client, will go to 10000 since it can't read the env:
```ts
        // codes in zustand
        const res = await isLogin({
          appid: process.env.NEXT_PUBLIC_BIZ_ENV === "global" ? 10001 : 10000,
        });
```
And I actually achieve the passing of `biz_env` by commands like this:
```js
let NEXT_PUBLIC_BIZ_ENV = "cn";
process.argv.forEach((arg) => {
  if (arg.startsWith("--biz-env=")) {
    NEXT_PUBLIC_BIZ_ENV = arg.split("=")[1];
  }
});

NODE_ENV=${NODE_ENV} --build-arg NEXT_PUBLIC_BIZ_ENV=${NEXT_PUBLIC_BIZ_ENV} -t ${dockerImageName} -f ./Dockerfile .`;

const dockerBuildCommand = `docker build --build-arg NEXT_PUBLIC_BIZ_ENV=${NEXT_PUBLIC_BIZ_ENV} -t ${dockerImageName} -f ./Dockerfile .`;
```
So I am wondering what is wrong. And also seems like by using `.env` or `.env.local` can work right, but I aslo want to know if there are some ways to change the env based on commands.

**Comment by Juneezee** [comment]
One thing you need to be aware of is Next.js inlines the environment variables at build time, see https://nextjs.org/docs/app/building-your-application/configuring/environment-variables#bundling-environment-variables-for-the-browser

> In order to make the value of an environment variable accessible in the browser, Next.js can "inline" a value, at build time, into the js bundle that is delivered to the client, replacing all references to `process.env.[variable]` with a hard-coded value.

**Reply by LikeDreamwalker** [reply]
Didn't get it. I'm using it like NEXT_PUBLIC_BIZ_ENV, and at build time I set it as 'global', and during the server time it will be 'global' as expected but not on the client.

I checked the document again and I am wondering shouldn't Next replace the judgment to true?Seems like process.env has been called again on the client, which uses a different env, and make sense.
